examples/pauseable.rs

```diff
@@ -0,0 +1,101 @@
+use std::thread;
+use std::time::{Duration, Instant};
+use winit::{
+    event::{ElementState, Event, KeyboardInput, VirtualKeyCode, WindowEvent},
+    event_loop::{ControlFlow, EventLoop},
+    window::WindowBuilder,
+};
+
+fn main() {
+    simple_logger::init().unwrap();
+    let event_loop = EventLoop::new();
+
+    let window = WindowBuilder::new()
+        .with_title("Press space to pause/resume")
+        .build(&event_loop)
+        .unwrap();
+
+    // This is an example of a pauseable game loop. While running == false, the application only
+    // responds to input events. When running == true, the application updates a stopwatch,
+    // tracking how much time has passed. The spacebar key pauses/resumes.
+    //
+    // TODO: Produce input events (by wiggling the mouse, for example). More than 10 updates per
+    // second happen.
+
+    // 10 updates per second
+    let update_frequency = Duration::from_millis(100);
+    let started_at = Instant::now();
+
+    let mut running = true;
+    let mut last_update = Instant::now();
+    let mut total = Duration::new(0, 0);
+
+    event_loop.run(move |event, _, control_flow| {
+        match event {
+            Event::WindowEvent {
+                event: WindowEvent::CloseRequested,
+                ..
+            } => *control_flow = ControlFlow::Exit,
+            Event::WindowEvent {
+                event:
+                    WindowEvent::KeyboardInput {
+                        input:
+                            KeyboardInput {
+                                virtual_keycode: Some(VirtualKeyCode::Space),
+                                state: ElementState::Pressed,
+                                ..
+                            },
+                        ..
+                    },
+                ..
+            } => {
+                window.request_redraw();
+                running = !running;
+                if running {
+                    let now = Instant::now();
+                    last_update = now;
+                    *control_flow = ControlFlow::WaitUntil(now + update_frequency);
+                    println!("- Resuming!");
+                } else {
+                    *control_flow = ControlFlow::Wait;
+                    println!("- Pausing");
+                }
+            }
+            Event::WindowEvent { .. } => {
+                println!("- Handling some input");
+                // Pretend that handling the event takes about 1ms
+                thread::sleep(Duration::from_millis(1));
+                window.request_redraw();
```

Comment:
calling `request_redraw` might result in undefined behavior
see https://docs.rs/winit/0.21.0/winit/window/struct.Window.html#method.request_redraw

best would be to call it in MainEventsCleared

Reply:
Done. Thanks for pointing out the caveat with request_redraw; it's a little unintuitive. I would expect to call it after changing any internal state that I want to be redrawn, and all of the requests during each WindowEvent to get batched together reasonably.